transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -802,22 +827,86 @@ private void write(Object msg, boolean flush, ChannelPromise promise) {
+            if (!(executor() instanceof SingleThreadEventLoop)) {
+                // The Executor is not of type SingleThreadEventLoop which means we have no way to schedule a
+                // flush at the end of the event loop run. Just explicitly do a flush in this case.
                 next.invokeWriteAndFlush(m, promise);
             } else {
-                next.invokeWrite(m, promise);
+                // Only schedule if there is no other flush pending yet.
+                if (!flushPending) {
+                    // First schedule it so if the scheduling throws any exception it will be propergated to
+                    // the promise and release te message.
+                    try {
+                        scheduleFlushOnEventLoopIteration();
+                    } catch (RejectedExecutionException e) {
+                        try {
+                            // If it was rejected but the promise is already completed we not care too much, otherwise
+                            // release the message.
+                            promise.setFailure(e);
+                        } finally {
+                            ReferenceCountUtil.release(m);
+                        }
+                        return;
+                    }
+                }
+                // We scheduled a flush so now one is pending.
+                next.invokeWrite(m, promise, true);
             }
         } else {
-            AbstractWriteTask task;
-            if (flush) {
-                task = WriteAndFlushTask.newInstance(next, m, promise);
-            }  else {
-                task = WriteTask.newInstance(next, m, promise);
+            next.invokeWrite(m, promise, false);
+        }
+    }
+
+    private void writeFromOutsideEventLoop(AbstractChannelHandlerContext next, EventExecutor executor, Object m,
+                                           boolean flush, ChannelPromise promise) {
+        if (flush) {
+            safeExecute(executor, WriteAndFlushTask.newInstance(next, m, promise, false), promise, m);
+        } else if (isAutoFlush()) {
+            if (!(executor() instanceof SingleThreadEventLoop)) {
+                // The Executor is not of type SingleThreadEventLoop which means we have no way to schedule a
+                // flush at the end of the event loop run. Just explicitly do a flush in this case.
+                safeExecute(executor, WriteAndFlushTask.newInstance(next, m, promise, true), promise, m);
+            } else {
+                // We first add the write task and then schedule something on the loop if needed. This is done
+                // so we are sure we not miss the flush. We will explicit handle the case of
+                // RejectedExecutionException in this case.
+                safeExecute(executor, WriteTask.newInstance(next, m, promise, true), promise, m);
+
+                // Only schedule if there is no other flush pending yet.
+                if (!flushPending) {
```

Comment:
I think it is ok as it is as at worse we will schedule multiple flushes.  The challenge here is really that if we would only keep track of the last Runnable we would need to keep track of where exactly in the pipeline this was triggered as we need to ensure we always execute the flush at least from the "last context" so all the expected handlers see the flush.